pkg/wireguard/agent/agent.go

```diff
@@ -0,0 +1,352 @@
+
+	a.Lock()
+	defer a.Unlock()
+
+	// Handle pubKey change
+	if prevPubKeyHex, found := a.pubKeyByNodeName[nodeName]; found && prevPubKeyHex != pubKeyHex {
+		log.WithField("nodeName", nodeName).Info("Pubkey has changed")
+		// pubKeys differ, so delete old peer
+		if err := a.deletePeerByPubKey(prevPubKeyHex); err != nil {
+			return err
+		}
+		delete(a.pubKeyByNodeName, nodeName)
+	}
+
+	log.WithFields(logrus.Fields{
+		"nodeName":  nodeName,
+		"pubKey":    pubKeyHex,
+		"nodeIPv4":  nodeIPv4,
+		"podCIDRv4": podCIDRv4,
+		"wgIPv4":    wgIPv4,
+		"nodeIPv6":  nodeIPv6,
+		"podCIDRv6": podCIDRv6,
+		"wgIPv6":    wgIPv6,
+	}).Info("Adding peer")
+
+	pubKey, err := wgtypes.ParseKey(pubKeyHex)
+	if err != nil {
+		return err
+	}
+
+	allowedIPs := []net.IPNet{}
+
+	if option.Config.EnableIPv4 {
+		if wgIPv4 != nil {
+			var peerIPNet net.IPNet
+			peerIPNet.IP = wgIPv4
+			peerIPNet.Mask = net.IPv4Mask(255, 255, 255, 255)
+			allowedIPs = append(allowedIPs, peerIPNet)
+		}
+		if podCIDRv4 != nil {
+			allowedIPs = append(allowedIPs, *podCIDRv4)
+		}
+	}
+	if option.Config.EnableIPv6 {
+		if wgIPv6 != nil {
+			var peerIPNet net.IPNet
+			peerIPNet.IP = wgIPv6
+			peerIPNet.Mask = net.CIDRMask(128, 128)
+			allowedIPs = append(allowedIPs, peerIPNet)
+		}
+		if podCIDRv6 != nil {
+			allowedIPs = append(allowedIPs, *podCIDRv6)
+		}
+	}
+
+	ep := ""
+	if option.Config.EnableIPv4 {
+		ep = net.JoinHostPort(nodeIPv4.String(), strconv.Itoa(listenPort))
+	} else if option.Config.EnableIPv6 {
+		ep = net.JoinHostPort(nodeIPv6.String(), strconv.Itoa(listenPort))
+	}
+	epAddr, err := net.ResolveUDPAddr("udp", ep)
+	if err != nil {
+		return err
+	}
+
+	peerConfig := wgtypes.PeerConfig{
+		Endpoint:          epAddr,
+		PublicKey:         pubKey,
+		AllowedIPs:        allowedIPs,
+		ReplaceAllowedIPs: true,
+	}
+	cfg := &wgtypes.Config{ReplacePeers: false, Peers: []wgtypes.PeerConfig{peerConfig}}
+	if err := a.wgClient.ConfigureDevice(IfaceName, *cfg); err != nil {
+		return err
+	}
+
+	a.pubKeyByNodeName[nodeName] = pubKeyHex
+
+	return nil
+}
+
+func (a *Agent) DeletePeer(nodeName string) error {
+	a.Lock()
+	defer a.Unlock()
+
+	pubKeyHex, found := a.pubKeyByNodeName[nodeName]
+	if !found {
+		return fmt.Errorf("cannot find pubkey for %s node", nodeName)
+	}
+
+	if err := a.deletePeerByPubKey(pubKeyHex); err != nil {
+		return err
+	}
+
+	delete(a.pubKeyByNodeName, nodeName)
+
+	return nil
+}
+
+func (a *Agent) deletePeerByPubKey(pubKeyHex string) error {
+	log.WithField("pubKey", pubKeyHex).Info("Removing peer")
+
+	pubKey, err := wgtypes.ParseKey(pubKeyHex)
+	if err != nil {
+		return err
+	}
+
+	peerCfg := wgtypes.PeerConfig{
+		PublicKey: pubKey,
+		Remove:    true,
+	}
+
+	cfg := &wgtypes.Config{Peers: []wgtypes.PeerConfig{peerCfg}}
+	if err := a.wgClient.ConfigureDevice(IfaceName, *cfg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
```

Comment:
Should we abstract this to use `os.Reader` instead of forcing the use of a file? It will ease unit tests

Reply:
Not sure how we can implement the body of this function with an `os.Reader`. It needs to check if the file exists or not. It should be unit-testable via  `t.TempDir`.